standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1201,22 +1199,86 @@ private TableValidWriteIds getValidWriteIdsForTable(Statement stmt, String fullT
+    List<TxnToWriteId> txnToWriteIds = rqst.getTxnToWriteIdList();
+    assert (txnToWriteIds != null);
+
+    // Get the next write id for the table and check if its in consistent with the input txntowriteids.
+    String s = sqlGenerator.addForUpdateClause(
+            "select nwi_next from NEXT_WRITE_ID where nwi_database = " + quoteString(dbName)
+                    + " and nwi_table = " + quoteString(tblName));
+    LOG.debug("Going to execute query <" + s + ">");
+    rs = stmt.executeQuery(s);
+    if (rs.next()) {
+      validWriteId = rs.getLong(1);
+    } else {
+      // All input write ids should be greater than this
+      validWriteId = 0L;
+    }
+
+    int size = txnToWriteIds.size();
+    for (TxnToWriteId txnToWriteId :  txnToWriteIds) {
+      Long targetTxnId = getTargetTxnId(rqst.getReplPolicy(), txnToWriteId.getTxnId(), stmt);
+      if (targetTxnId == -1) {
+        LOG.info("Transaction " + txnToWriteId.getTxnId() +
+                " is invalid, may be already committed/aborted. Ignoring the whole operation");
+        return new ArrayList<>();
+      }
+
+      Long writeId = txnToWriteId.getWriteId();
+      if (writeId <= validWriteId) {
+        //looks like a retry, so do a noop
+        LOG.info("Write Id " + writeId + " is less than valid write id : " + validWriteId + "Looks like retry.");
+        return new ArrayList<>();
+      }
+
+      if (writeId > (validWriteId + size)) {
+        //this should never happen in case of replication
+        LOG.info("Write Id " + writeId + " is out of range. Valid write id : " + validWriteId + " Size : " + size);
+        throw new MetaException("Write Id " + writeId + " is out of range. Valid write id : " +
+                validWriteId + " Size : " + size);
+      }
+      txnIds.add(targetTxnId);
+    }
+
+    return txnIds;
+  }
+
   @Override
+  @RetrySemantics.Idempotent
   public AllocateTableWriteIdsResponse allocateTableWriteIds(AllocateTableWriteIdsRequest rqst)
           throws NoSuchTxnException, TxnAbortedException, MetaException {
-    List<Long> txnIds = rqst.getTxnIds();
+    List<Long> txnIds;
     String dbName = rqst.getDbName().toLowerCase();
     String tblName = rqst.getTableName().toLowerCase();
     try {
       Connection dbConn = null;
       Statement stmt = null;
       ResultSet rs = null;
       TxnStore.MutexAPI.LockHandle handle = null;
+      List<TxnToWriteId> txnToWriteIds = new ArrayList<>();
       try {
         lockInternal();
         dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
         stmt = dbConn.createStatement();
 
+        if (rqst.isSetReplPolicy()) {
+          txnIds = getValidTxnIds(rqst, stmt, rs);
+        } else {
+          txnIds = rqst.getTxnIds();
+        }
+
+        if (txnIds.size() == 0) {
+          // Idempotent case for replication flow, return the same txn2writeId list as in request message.
+          return new AllocateTableWriteIdsResponse(rqst.getTxnToWriteIdList());
```

Comment:
Can't return the input txnToWriteIds as they are mapped against source txn id. Better to return empty list in this case to denote no allocation happened but didn't fail as it is repl flow.